### ISSUE TITLE
Support for conditional/hierarchical design spaces

### ISSUE THREAD

**Issue by bbudescu** [huawei-noah/HEBO#30]
Can one specify a conditional/hierarchical search space to HEBO? I.e., something similar to [SMAC3](https://automl.github.io/ConfigSpace/main/guide.html#nd-example-categorical-hyperparameters-and-conditions)?

E.g., only sample the number of convoutional filters in the second layer of a neural net only if we decide (by another parameter )to have a network with at least two layers.

I'm thinking that this can be somewhat circumvented by returning a high cost value for infeasible combinations, but I imagine this might be a suboptimal approach as it might affect optimization performance since, e.g., some optima might lie on the edge of feasible regions, and the cost estimator, if it has some sort of  smoothness prior (which they usually do), has a chance of assigning unfaithfully high cost values near the infeasible configurations - at least initially (i.e., the probability of a kink in the error surface is generally lower).

[Some optimizers](https://github.com/luinardi/hypermapper/wiki/Chakong%E2%80%90Haimes) are able to address this by training a different model - a feasability predictor (which has the advantage of being able to work with unknown feasibility constraints).

So how can one

**Comment by artiko88** [comment]
Has anyone come up with any solution for hierarchical searches using HEBO? Would hate to switch to different optimizer because of not being able to do something similar to how SMAC3 performs hierarchical searches.
